### PR TITLE
Implement moveToApplicationsFolder

### LINKED ISSUE #1692: App should offer to move itself to /Applications if run from elsewhere
This would be a very nice UX feature. Many users would just run the app directly from the `Downloads` directory.

A lot of apps use this module to achieve it: https://github.com/potionfactory/LetsMove

I realize this is a longshot, but I really like apps that do this and would be awesome if all Electron apps could easily do it.

image

### PR DESCRIPTION
**Pull request by MarshallOfSound**
Closes #1692

This adds two methods:

* `app.isInApplicationsFolder()` - Is the app currently in the system applications folder
* `app.moveToApplicationsFolder()` - Attempt to move the app to the Applications folder and relaunch on success

Any UX for asking the user if they want to move the app should be implemented by the app either through HTML/CSS or using the `dialog` module.

### CODE REVIEW

**Comment by MarshallOfSound** [comment]
Please note this is heavily influenced (and in fact mostly transposed) by [LetsMove](https://github.com/potionfactory/LetsMove/blob/master/PFMoveApplication.m)

File: docs/api/app.md [hunk 1]
```diff
@@ -927,6 +927,27 @@ Enables mixed sandbox mode on the app.
...
 
+### `app.isInApplicationsFolder()` _macOS_
+
+Returns `Boolean` - Whether the application is currently running from the
+systems Application folder.  Use in combination with `app.moveToApplicationsFolder()`
+
+### `app.moveToApplicationsFolder()` _macOS_
+
+Returns `Boolean` - Whether the move was successful.  Please note that if
+the move was successful your application will be quit and relaunched
```
Comment (enlight) [inline docs/api/app.md]: -> "if the move is successful your application will quit and relaunch"

File: docs/api/app.md [hunk 2]
```diff
@@ -927,6 +927,27 @@ Enables mixed sandbox mode on the app.
...
+automatically in it's new location.
+
+This method does not have any dialog or confirmation by default, if you want
+to ask your user if you should make this move then you should use the `dialog`
+API's.
```
Comment (enlight) [inline docs/api/app.md]: -> No confirmation dialog will be presented by default, if you wish to allow the user to confirm the operation you may do so using the `dialog` API.

**Comment by jkleinsc** [comment]
@MarshallOfSound while this is useful functionality, it is a lot of code to add to Electron for a very specific use case.  It seems like it would be better to do this via a library/package such as [electron-lets-move](https://www.npmjs.com/package/electron-lets-move).

**Comment by baconbrad** [comment]
I second @jkleinsc's remark. If it is a particular and completely optional use case that can be done as an NPM module then I don't see why that shouldn't be the recommended way. While this code doesn't bloat Electron by itself it would add to the overall bloat in Electron. Additionally those who use auto updaters in their applications that bandwidth from overall bloat starts to add up overtime as well.

**Comment by MarshallOfSound** [comment]
@jkleinsc I'm not sure if we have a policy on core / module but the way I look at it, it isn't just a question of whether this *can* be implemented in a module (I'm pretty sure it can be).  It's a question of developer friendliness and also user friendliness (people using Electron apps).

If we implement this in core more developers will use it and therefore users would get a better experience when using Electron apps.  I'm perfectly happy to move this into a module if that's what we want to do.

In terms of risk from the amount of code that is added, this code has been battle tested for years in the LetsMove repository so the code itself comes with minimal risk (probably less than most additions to core).

**Comment by zeke** [comment]
What about moving it to a module that is then depended on by core? Is there a precedent for that?

**Comment by enlight** [comment]
I'm in the keep the core minimal camp, if something can be implemented in user-land without jumping through too many hoops then it should be. Separate modules also make it possible for users to upgrade/downgrade a module while staying on the same Electron version (in some cases), and to customize modules without maintaining a fork of Electron.

@MarshallOfSound certainly has a point about people getting tripped up by native Node modules ALL THE TIME, but that needs to be addressed by some kind of on-demand native module build bot.

**Comment by poiru** [comment]
I agree that we shouldn't have niche features in core, but this would useful to almost all macOS apps. Is it worth the complexity of having this as a separate module? The binary size impact of this is probably a couple kilobytes at most. Having this as a separate module would almost certainly result in a net size increase.

Compare this to e.g. PDF support or WebRTC, which are in the megabytes. If we want to reduce bloat, we should start with the big offerenders.

File: docs/api/app.md [hunk 3]
```diff
@@ -927,6 +927,26 @@ Enables mixed sandbox mode on the app.
...
+
+Returns `Boolean` - Whether the move was successful.  Please note that if
+the move is successful your application will quit and relaunch.
+
+No confirmation dialog will be presented by default, if you wish to allow
+the user to confirm the operation you may do so using the
+[`dialog`](dialog.md) API.
+
+**NOTE:** This method throws errors if anything other than the user causes the
+move to fail.  For instance if the user cancels the authorization dialog this
```
Comment (zeke) [inline docs/api/app.md]: Under what condition does this method return `false`? That is, when would it not be "successful" but also not throw?

Comment (MarshallOfSound) [inline docs/api/app.md]: If the move requires elevated permissions and the user cancels the elevated permissions prompt this method returns false.

**Comment by zeke** [comment]
> Is it worth the complexity of having this as a separate module?

I don't know, but I'm curious. I just want to be sure we're not missing opportunities to modularize Electron's codebase.

**Comment by zeke** [comment]
cc @tommoor who wrote https://github.com/tommoor/electron-lets-move and might have some insight.

**Comment by davej** [comment]
+1 to add this in core. `electron-lets-move` is great but it relies on [`sudo-prompt`](https://github.com/jorangreef/sudo-prompt) which is a bit of a hack.  Basically what [`sudo-prompt`](https://github.com/jorangreef/sudo-prompt) does is:

> 1. Use "which" to lookup the location of the AppleScript binary (osascript) and copy this to a tmp path, but using the name of your application as the name of the copied binary. This causes the OS X permissions prompt to show the name of your application instead of the AppleScript binary.
> 2. Use the new AppleScript binary to set the sudo timestamp for your user:
> 
> ```
> var command = '"' + tmposascript + '" -e \'do shell script "mkdir -p /var/db/sudo/$USER; touch /var/db/sudo/$USER" with administrator privileges\'';
> ```
>
> 3. This command will cause OS X to show a UI to the user asking for their password.
>
> 4. Finally, re-run the command that required privileges.

(from https://github.com/electron/electron/issues/1692#issuecomment-109974222)

I think it would be much cleaner to do this natively in node.

This leads to another issue, is it worth having an API for running a command with permission? Perhaps something like `shell.authorizedChildProcess` which would pop up a permissions GUI dialog and give access to a [`child_process`](https://nodejs.org/api/child_process.html) with permissions of the authenticated user? Or even something a bit more basic`shell.authorizedExec` which would just allow `exec`?

Let me know if it's worthwhile to create an issue for the above API?

**Comment by jkleinsc** [comment]
@MarshallOfSound given the feedback above and @zeke's decision to approve this, I think we should move this forward.  It looks like some of the tests are failing (linux in particular).  Can you take a look at the failing tests?

**Comment by MarshallOfSound** [comment]
@jkleinsc Looks like the spec was flawed, should have only been running on darwin.  Fixed now, the other failures were just flakes IMO

Once these builds run we should be gtg

**Comment by jkleinsc** [comment]
@MarshallOfSound thanks for updating.  It looks like there is an extra semicolon in the test:
  /Users/electron/workspace/electron-mas-x64/spec/api-app-spec.js:119:48: Extra semicolon.

**Comment by MarshallOfSound** [comment]
:rip: Always mess up switching between airbnb and standard JS 😆

**Comment by jkleinsc** [comment]
Looks good to me now.  I'll merge it in.

**Comment by djalmaaraujo** [comment]
@buzinas @philfreo ☝️

**Pull request merged**